packages/xl-ai/src/prosemirror/agent.ts

```diff
@@ -0,0 +1,325 @@
+            newValue: newNode.type.name,
+          })
+          .addToSet(marks);
+      }
+
+      const attrNames = new Set([
+        ...Object.keys(newNode.attrs),
+        ...Object.keys(oldNode.attrs),
+      ]);
+      for (const attr of attrNames) {
+        if (newNode.attrs[attr] !== oldNode.attrs[attr]) {
+          marks = modification
+            .create({
+              type: "attr",
+              attrName: attr,
+              previousValue: oldNode.attrs[attr],
+              newValue: newNode.attrs[attr],
+            })
+            .addToSet(marks);
+        }
+      }
+
+      const stepIndex = tr.steps.length;
+      tr.setNodeMarkup(
+        tr.mapping.map(step.from),
+        newNode.type,
+        newNode.attrs,
+        marks,
+      );
+
+      agentSteps.push({
+        prosemirrorSteps: tr.steps.slice(stepIndex),
+        selection: undefined,
+        type: "replace",
+      });
+
+      continue;
+    }
+
+    if (!(step instanceof ReplaceStep)) {
+      throw new Error("Step is not a ReplaceStep");
+    }
+
+    if (step.slice.openStart > 0 || step.slice.openEnd > 0) {
+      // these kind of changes should not be generated by changeset.ts
+      throw new Error(
+        "Slice has openStart or openEnd > 0, but structure=false",
+      );
+    }
+
+    // 1. Select text to be removed/replaced
+    agentSteps.push({
+      prosemirrorSteps: [],
+      selection: {
+        anchor: tr.mapping.map(step.from),
+        head: tr.mapping.map(step.to),
+      },
+      type: "select",
+    });
+
+    // Replace the content
+    const sliceTextContent = step.slice.content.textBetween(0, step.slice.size);
+
+    const alreadyHasSameText =
+      sliceTextContent ===
+      tr.doc.textBetween(tr.mapping.map(step.from), tr.mapping.map(step.to));
+
+    let sliceTo: number;
+
+    if (alreadyHasSameText) {
+      sliceTo = step.slice.content.size; // replace all at once, it's probably a mark update
+    } else if (sliceTextContent.length === 0) {
+      sliceTo = step.slice.content.size; // there's no replacement text, so use entire slice as replacement
+    } else {
+      // replace with the first character (similar to how a user would do it when selecting text and starting to type)
+      const firstCharIndex = getFirstChar(step.slice.content);
+      if (firstCharIndex === undefined) {
+        // should have been caught by previous if statement
+        throw new Error("unexpected: no first character found");
+      }
+      sliceTo = firstCharIndex + 1;
+    }
+
+    let replaceEnd = tr.mapping.map(step.to);
+    const replaceFrom = tr.mapping.map(step.to);
+    let first = true;
+
+    // Note that below, we don't actually delete content, but we mark it as deleted
+    // with a deletion mark. Similarly, we mark inserted content with an insertion mark.
+    // It might be cleaner to;
+    // a) make this optional
+    // b) actually delete / insert the content and let prosemirror-suggest-changes handle the marks
```

Comment:
Yea, I would consider doing b, it feels like prosemirror-suggest-changes should be doing this transformation, though I can understand wanting the fine-grained control of how it displays visually.

Reply:
Yeah the initial prototype used the current approach, I only later hooked up prosemirror-suggest-changes. I tried to migrate to having that handle the insertions, but was running into a few edge-cases which made it more cumbersome than I initially hoped (ofc, might have just been a single mapping / positioning error). For now I think we can keep it as is, but good to leave the note here if we run into things that the "other approach" could address, wdyt?

Reply:
Totally fine with that. If we experience a bug with it, I'd probably just throw it away and re-implement